File: C-Plus-Plus/shell_sort.cpp
#include <bits/stdc++.h>

using namespace std;

void shellSort(int n, vector<int> arr) {
    for (int dif = n/2; dif > 0; dif /= 2) {
        for (int i = dif; i < n; ++i) {
            int temp = arr[i];
            int j;
            for (j = i; j >= dif && arr[j - dif] > temp; j -= dif) {
                arr[j] = arr[j - dif];
            }
            arr[j] = temp;
        }
    }
    cout << "Sorted array: ";
    for (int i = 0; i < n; ++i) {
        cout << arr[i] << " ";
    }
    
}

int main() {
    int n;
    cout << "Enter the number of elements in array:";
    cin >> n;
    vector<int> arr(n);
    cout << "Enter the elements:" << endl;
    for (int i = 0; i < n; ++i) {
        cin >> arr[i];
    }

    shellSort(n, arr);
    
    return 0;
}

/*
 * Sample input:
 * Enter the number of elements in array:5
 * Enter the elements:
 * 12 34 54 2 3
 * Output:
 * Sorted array: 2 3 12 34 54
 *
 * Time Complexity: O(n^2)
 * Space Complexity: O(1)
 */

File: Java/Sorting/ShellSort.java
package Java.Sorting;

/**
 ** @author Nathan Nasteff
 */

public class ShellSort {

	/* 
	** The shell sort algorithm uses a directed insersion sort
	** to first sort widely spaced array elements (length / 2)
	** and then elements that are closer together (length /= 2)
	*/

	public static void shellSort(int arr[]) {

		int a = arr.length;

		for (int interval = a / 2; interval > 0; interval /= 2) {

			for (int i = interval; i < a; i += 1) {

				int temp = arr[i];
				int j;

				for (j = i; j >=  interval && arr[j - interval] > temp; j -= interval)
					arr[j] = arr[j - interval];

				arr[j] = temp;
			}
		}
	}

	public static void main(String[] args) {

		// Test case

		int test[] = new int[] { 30, 10, 12, 1, 4, 6, 2, 19 };
		int i;

		System.out.println("Original Array:");
		for (i = 0; i < test.length; i++) {
			System.out.print(test[i] + " ");
			if (i == (test.length - 1)) {
				System.out.println("\n");
			}
		}

		shellSort(test);

		System.out.println("Sorted Array:");
		for (i = 0; i < test.length; i++) {
			System.out.print(test[i] + " ");
		}
	}
}

/*
** Sample I/O
**
** INPUT:
** 
** 30 10 12 1 4 6 2 19
**
** OUTPUT:
**
** Original Array:
** 30 10 12 1 4 6 2 19
** 
** Sorted Array:
** 1 2 4 6 10 12 19 30
**
** Time Complexity - O(n log n) 
** Space Complexity - O(1)
**
*/
